### DISCUSSION TITLE
How to implement Trainer.predict to save test images

### DISCUSSION THREAD

**Discussion by bibinwils** [Lightning-AI/pytorch-lightning#9278] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Hi, I trained a UNet with input images and mask. For that, I split the entire dataset to train - valid - test. I have trained the network and got the output metrics for the test dataset. There are around 100 test images. Now I would like to save all the test data outputs to a folder with the same name as the test images. For example, if the test image name is image1, then I need to save the test output image also as image1. I have gone through the PyTorch lightning documentation, I come to know about Trainer.predict, but I don't know how to implement it.

**Comment by pederismo** [comment]
You have to implement the predict_step(...) function in your LightningModule. Now inside of this function you can torch.save(...) your outputs (which I assume are going to be segmentations?). In order to have the name of the output file to be the same of the input file, my suggestion would be to include the name of the input file in the batch (if you've defined your own DataLoader), something like this:

```
def __getitem__(index):
    ...
    batch = {"image": image, "label": label, "filename": filename}
    return batch
```

**Reply by bibinwils** [reply]
Can you please elaborate. I'm new to this. I couldn't find any predict_step() in the documentation. also how to call the filename from data loader as you have mentioned to the predict_step()

**Reply by bibinwils** [reply]
anybody here to help?

**Reply by rohitgr7** [reply]
here is the doc for [predict_step](https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#predict-step)

to make it work you need to either define [predict_dataloader](https://pytorch-lightning.readthedocs.io/en/stable/extensions/datamodules.html#predict-dataloader) or pass a dataloader inside [trainer.predict](https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#predict-step),

**Reply by bibinwils** [reply]
Thanks. I tried so. Now I'm getting the following error

IndexError: dimension specified as 0 but tensor has no dimensions
![image](https://user-images.githubusercontent.com/53151992/132549673-7b51da48-ee97-4a92-b8de-b0cf047c0c39.png)
